File: tests/sdk-smoke-python/negotiated_continuation.py
from __future__ import annotations

import os
import sys
from pathlib import Path
from urllib.parse import urlsplit

import openai

sys.path.insert(0, str(Path(__file__).resolve().parents[2] / "clients/continuation"))
from olp_continuation import next_turn, recover_submission, stream_turn, unary_turn  # noqa: E402

origin = os.environ["OLP_CONTINUATION_ORIGIN"]
key = os.environ["OLP_CONTINUATION_KEY"]
route = os.environ["OLP_CONTINUATION_ROUTE"]
parsed = urlsplit(origin)
assert parsed.scheme == "http" and parsed.hostname == "127.0.0.1" and parsed.port
assert key.startswith("olp_") and route.startswith("strict-")

class UnsupportedClient:
    def __init__(self) -> None:
        self.calls = 0

    @property
    def chat(self) -> "UnsupportedClient":
        self.calls += 1
        return self

unsupported = UnsupportedClient()
try:
    stream_turn(unsupported, {"model": route, "messages": []})
    raise AssertionError("unsupported SDK unexpectedly dispatched")
except ValueError as error:
    assert "supports the qualified" in str(error)
assert unsupported.calls == 0

request = {
    "model": route,
    "stream": True,
    "tools": [
        {"type": "function", "function": {"name": "weather", "description": "Weather in a city", "parameters": {"type": "object", "properties": {"city": {"type": "string"}}, "required": ["city"]}}},
        {"type": "function", "function": {"name": "clock", "description": "Time in a zone", "parameters": {"type": "object", "properties": {"zone": {"type": "string"}}, "required": ["zone"]}}},
    ],
    "messages": [{"role": "user", "content": "Weather and time in Paris?"}],
}
with openai.OpenAI(api_key=key, base_url=f"{origin}/v1", max_retries=0, timeout=15.0) as client:
    try:
        client.chat.completions.create(**request)
        raise AssertionError("unnegotiated SDK unexpectedly dispatched")
    except openai.BadRequestError as error:
        assert error.status_code == 400
        assert error.code == "state_carrier"

    first = stream_turn(client, request)
    assert first["version"] == "chat-anthropic-tools-v1"
    assert first["finish"] == "tool_calls"
    # F01 regression: the unified completed-turn value exposes the canonical
    # plain assistant mapping, never the raw SDK message object.
    assert isinstance(first["assistant"], dict)
    assert first["assistant"]["content"] == "beforeafter"
    assert [call["id"] for call in first["assistant"]["tool_calls"]] == ["call-weather", "call-clock"]
    assert first["tools"] == first["assistant"]["tool_calls"]
    assert first["actions"] == {"tool_calls": ["call-weather", "call-clock"]}
    assert [item["type"] for item in first["observations"] if item["phase"] == "start"] == [
        "thinking", "text", "tool_use", "tool_use", "text"
    ]
    assert any(item["type"] == "thinking" and item.get("opaque_state") is True for item in first["observations"])
    assert first["native_usage"] == {"input_tokens": 18, "output_tokens": 28}
    assert first["native_terminal"] == {"stop_reason": "tool_use", "stop_sequence": None, "finish_reason": "tool_calls"}
    assert "opaque-fixture-signature-do-not-log" not in repr(first["chunks"])
    recovered = recover_submission(origin, key, first["submission"])
    assert recovered["handle"] == first["handle"]
    assert recovered["assistant"] == first["assistant"]
    assert recovered["native_terminal"] == first["native_terminal"]
    assert recovered["actions"] == first["actions"]
    assert isinstance(recovered["delivery"]["frames"], list) and recovered["delivery"]["frames"]
    replay = stream_turn(client, request, submission=first["submission"])
    assert replay["handle"] == first["handle"]
    assert replay["assistant"] == first["assistant"]
    # The explicit claim, not the ready handle alone, gates tool actions.
    for results in (
        [{"tool_call_id": "call-clock", "content": "14:00"}, {"tool_call_id": "call-weather", "content": "sunny"}],
        [{"tool_call_id": "call-weather", "content": "sunny"}],
    ):
        try:
            next_turn(request, first, results)
            raise AssertionError("non-corresponding tool results accepted")
        except ValueError as error:
            assert "call order and identity" in str(error) or "each" in str(error)
    try:
        next_turn(request, {**first, "actions": None}, [
            {"tool_call_id": "call-weather", "content": "sunny"},
            {"tool_call_id": "call-clock", "content": "14:00"},
        ])
        raise AssertionError("handle without an action claim yielded calls")
    except ValueError as error:
        assert "actionability" in str(error)
    next_request = next_turn(request, first, [
        {"tool_call_id": "call-weather", "content": "sunny"},
        {"tool_call_id": "call-clock", "content": "14:00"},
    ])
    final = unary_turn(client, next_request, handle=first["handle"])
    assert final["assistant"]["content"] == "Both tools completed."
    assert final["finish"] == "stop"
    assert final["actions"] == {"tool_calls": []}
    try:
        next_turn(next_request, final, [])
        raise AssertionError("a completed turn yielded tool actions")
    except ValueError as error:
        assert "no tool actions" in str(error)
    assert final["native_usage"] == {"input_tokens": 30, "output_tokens": 4}
    assert final["native_terminal"] == {"stop_reason": "end_turn", "stop_sequence": None, "finish_reason": "stop"}
    print(f"openai-python-3.8.0: {len(first['observations'])} observations; two native dispatches")

File: tests/sdk-smoke/negotiated-continuation.mjs
import assert from 'node:assert/strict';
import OpenAI from 'openai';
import { nextTurn, recoverSubmission, streamTurn, unaryTurn } from '../../clients/continuation/javascript.mjs';

const { OLP_CONTINUATION_ORIGIN: origin, OLP_CONTINUATION_KEY: key, OLP_CONTINUATION_ROUTE: route } = process.env;
assert.match(origin, /^http:\/\/127\.0\.0\.1:\d+$/);
assert.ok(key?.startsWith('olp_'));
assert.ok(route?.startsWith('strict-'));
const local = globalThis.fetch.bind(globalThis);
const localOnlyFetch = (input, init) => {
  assert.equal(new URL(input instanceof Request ? input.url : String(input)).origin, origin);
  return local(input, init);
};
const client = new OpenAI({
  apiKey: key,
  baseURL: `${origin}/v1`,
  maxRetries: 0,
  timeout: 15_000,
  fetch: localOnlyFetch
});
let unsupportedCalls = 0;
await assert.rejects(() => streamTurn({
  getUserAgent: () => 'OpenAI/JS 0.0.0',
  chat: { completions: { create: () => { unsupportedCalls++; } } }
}, { model: route, messages: [] }), /supports the qualified/);
assert.equal(unsupportedCalls, 0, 'unsupported SDK rejected before dispatch');
const request = {
  model: route,
  stream: true,
  tools: [
    { type: 'function', function: { name: 'weather', description: 'Weather in a city', parameters: { type: 'object', properties: { city: { type: 'string' } }, required: ['city'] } } },
    { type: 'function', function: { name: 'clock', description: 'Time in a zone', parameters: { type: 'object', properties: { zone: { type: 'string' } }, required: ['zone'] } } }
  ],
  messages: [{ role: 'user', content: 'Weather and time in Paris?' }]
};
try {
  await client.chat.completions.create(request);
  assert.fail('unnegotiated SDK unexpectedly dispatched');
} catch (error) {
  assert.equal(error.status, 400);
  assert.equal(error.code, 'state_carrier');
}
const first = await streamTurn(client, request);
assert.equal(first.version, 'chat-anthropic-tools-v1');
assert.equal(first.finish, 'tool_calls');
// F01 regression: the unified completed-turn value exposes the canonical
// plain assistant mapping, never the raw SDK message object.
assert.equal(Object.getPrototypeOf(first.assistant), Object.prototype);
assert.equal(first.assistant.content, 'beforeafter');
assert.deepEqual(first.assistant.tool_calls.map((call) => call.id), ['call-weather', 'call-clock']);
assert.deepEqual(first.tools, first.assistant.tool_calls);
assert.deepEqual(first.actions, { tool_calls: ['call-weather', 'call-clock'] });
assert.deepEqual(first.observations.filter((item) => item.phase === 'start').map((item) => item.type),
  ['thinking', 'text', 'tool_use', 'tool_use', 'text']);
assert.ok(first.observations.some((item) => item.type === 'thinking' && item.opaque_state === true));
assert.deepEqual(first.nativeUsage, { input_tokens: 18, output_tokens: 28 });
assert.deepEqual(first.nativeTerminal, { stop_reason: 'tool_use', stop_sequence: null, finish_reason: 'tool_calls' });
assert.ok(!JSON.stringify(first.chunks).includes('opaque-fixture-signature-do-not-log'));
const recovered = await recoverSubmission(origin, key, first.submission, localOnlyFetch);
assert.equal(recovered.handle, first.handle);
assert.deepEqual(recovered.assistant, first.assistant);
assert.deepEqual(recovered.nativeTerminal, first.nativeTerminal);
assert.deepEqual(recovered.actions, first.actions);
assert.ok(Array.isArray(recovered.delivery.frames) && recovered.delivery.frames.length > 0);
const replay = await streamTurn(client, request, { submission: first.submission });
assert.equal(replay.handle, first.handle);
assert.deepEqual(replay.assistant, first.assistant);
// The explicit claim, not the ready handle alone, gates tool actions.
assert.throws(() => nextTurn(request, first, [
  { tool_call_id: 'call-clock', content: '14:00' },
  { tool_call_id: 'call-weather', content: 'sunny' }
]), /match call order and identity/);
assert.throws(() => nextTurn(request, first, [
  { tool_call_id: 'call-weather', content: 'sunny' }
]), /one result for each/);
assert.throws(() => nextTurn(request, { ...first, actions: undefined }, [
  { tool_call_id: 'call-weather', content: 'sunny' },
  { tool_call_id: 'call-clock', content: '14:00' }
]), /no actionability claim/);
const next = nextTurn(request, first, [
  { tool_call_id: 'call-weather', content: 'sunny' },
  { tool_call_id: 'call-clock', content: '14:00' }
]);
const final = await unaryTurn(client, next, { handle: first.handle });
assert.equal(final.assistant.content, 'Both tools completed.');
assert.equal(final.finish, 'stop');
assert.deepEqual(final.actions, { tool_calls: [] });
assert.throws(() => nextTurn(next, final, []), /no tool actions/);
assert.deepEqual(final.nativeUsage, { input_tokens: 30, output_tokens: 4 });
assert.deepEqual(final.nativeTerminal, { stop_reason: 'end_turn', stop_sequence: null, finish_reason: 'stop' });
process.stdout.write(JSON.stringify({ sdk: 'openai-js-7.4.0', first: first.handle, final: final.handle, observations: first.observations.length }) + '\n');
